### DISCUSSION TITLE
Calling PSCmdlets based on the instance instead of the name

### DISCUSSION THREAD

**Discussion by glatzert** [PowerShell/PowerShell#19493] [Q&A] [answered]
I'm currently stuck with a problem in a module, I try to use.
I could work around that problem, if I'd be able to call the PSCmdLet from that module (`Microsoft.IdentityServer.Deployment.Commands.InstallFarmCommand` to be precise).
That module has Switches, that are not offered via PowerShell, but are present as classic properties, so I could just modify them and finish my task. Unfortunately there does not seem to be a mechanism to call an instantiiated PSCmdLet?

So my Question: How can I add an already instantiated PSCmdLet to a pipeline to execute it? OR: What do I have to provide, so the PSCmdLet will allow Invoke()?

**Comment by kasini3000** [comment]
You seem to be looking for a way to multithread, if so, please take a look:
1       $r2 = [powershell]::Create().AddScript('#powershell code').beginInvoke() 

---

2      get-content ip.txt |foreach-object -Parallel { ping.exe $_  }
At present, there is a drawback, which is that all tasks have timed out.

**Reply by glatzert** [reply]
I just want to invoke the _instance_ of the PSCmdLet. I don't care about multithreading or something else.

**Comment by doctordns** [comment]
What are you trying to achieve = what is the use-case?

**Reply by glatzert** [reply]
Im trying to run the command with the _property_ (not _Parameter_) "SkipPreChecks" set to `$true`.
Therefore I (probably) need to instantiate the PSCmdLet and run it after I set that property.

```
$ic = [Microsoft.IdentityServer.Deployment.Commands.InstallFarmCommand]::new();
$ic.FederationServiceName = "login.contoso.net";
$ic.CertificateThumbprint = "71DF6135555585555D67555554BD50B0E90DAD0";
$ic.GroupServiceAccountIdentifier = "CONTOSO\srv_adfs";
$ic.AdminConfiguration=@{"DKMContainerDn"="CN=<guid>,CN=ADFS,CN=Microsoft,CN=Program Data,DC=CONTOSO,DC=Net"}; 
$ic.SkipPreChecks = $true # this is the relevant line of code, where the property is NOT marked as Parameter and thus cannot be set inline.

**Comment by doctordns** [comment]
I am not really aware of a method of calling this cmdlet (I assume you are trying to invoke Install-ADFSFarm) and specifying anything other than the parameters which seem to exclude one to skip any pre-checks. And would such use, even if we do know how, actually work? I can not find any documentation that discusses skipping prereq tests, but that could be weak Search Engine foo.

**Comment by glatzert** [comment]
I'm deep into debugging installing it from an underprivileged account (for academic reasons)... DotPeek knew about that switch.

Am 12.04.2023 16:52 schrieb Thomas Lee ***@***.***>:

I am not really aware of a method of calling this cmdlet (I assume you are trying to invoke Install-ADFSFarm) and specifying anything other than the parameters which seem to skip any pre-checks. And would such use, even if we do know how, actually work? I can not find any documentation that discusses skipping prereq tests, but that could be weak Search Engine foo.

**Comment by SeeminglyScience** [comment, marked as answer]
There is no way to invoke an instance of `PSCmdlet` without the pipeline. The engine must create the instance and populate it with the required state. The only supported way to do this is via a PowerShell pipeline.

If the switch is not surfaced as a parameter then it's not accessible, I'd suggest filing some feedback with the module owner.

(note that the `Invoke` method you reference only works if the command **only** inherits `Cmdlet` directly. If `PSCmdlet` is inherited, then it can no longer be invoked outside of a pipeline)
